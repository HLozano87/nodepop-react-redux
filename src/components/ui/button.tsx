import clsx from "clsx";

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: "primary" | "secondary";
}

export const Button = ({
  children,
  className,
  variant,
  disabled,
  onClick,
  ...props
}: ButtonProps) => {
  return (
    <button
      className={clsx(`btn-${variant}`, className, {
        "cursor-not-allowed opacity-50": disabled,
        "cursor-pointer": !disabled,
      })}
      disabled={disabled}
      onClick={onClick}
      {...props}
    >
      {children}
    </button>
  );
};
