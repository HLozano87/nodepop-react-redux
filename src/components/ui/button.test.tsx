import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Button } from "./button";

describe("rendering button with snapshot", () => {
  test("matches snapshot", () => {
    const { asFragment } = render(<Button>Click me</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Button", () => {
  test("renders children text", () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText("Click me")).toBeInTheDocument();
  });

  test("applies the correct variant class", () => {
    const { rerender } = render(<Button variant="primary">Primary</Button>);
    expect(screen.getByRole("button")).toHaveClass("btn-primary");

    rerender(<Button type="submit" variant="secondary">Secondary</Button>);
    expect(screen.getByRole("button")).toHaveClass("btn-secondary");
  });

  test("applies disabled classes and disables the button", () => {
    render(<Button disabled>Disabled</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
    expect(button).toHaveClass("cursor-not-allowed", "opacity-50");
    expect(button).not.toHaveClass("cursor-pointer");
  });

  test("applies cursor-pointer class when enabled", () => {
    render(<Button>Enabled</Button>);
    const button = screen.getByRole("button");
    expect(button).not.toBeDisabled();
    expect(button).toHaveClass("cursor-pointer");
    expect(button).not.toHaveClass("cursor-not-allowed");
  });

  test("calls onClick when clicked and not disabled", () => {
    const onClick = vi.fn();
    render(<Button onClick={onClick}>Click me</Button>);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test("does not call onClick when disabled", () => {
    const onClick = vi.fn();
    render(
      <Button disabled onClick={onClick}>
        Click me
      </Button>,
    );
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(onClick).not.toHaveBeenCalled();
  });

  test("merges additional className prop", () => {
    render(
      <Button className="my-custom-class" variant="primary">
        Custom Class
      </Button>,
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("btn-primary");
    expect(button).toHaveClass("my-custom-class");
  });
});
