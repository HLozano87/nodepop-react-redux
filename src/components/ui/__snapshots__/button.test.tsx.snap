// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rendering button with snapshot > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn-undefined cursor-pointer"
  >
    Click me
  </button>
</DocumentFragment>
`;
